### ISSUE TITLE
NUnit 3 Test Adaptor (3.9.0): Numerous InvalidOperationException from Mono.Cecil and no tests in Test Explorer

### ISSUE THREAD

**Issue by stephenmartindale** [nunit/nunit#2708]
A new problem spontaneously started occurring, this morning, with my solution that uses NUnit 3.9.0 and NUnit Test Adaptor for Visual Studio (also 3.9.0). So far, I have found no work-around.

Last Friday, all was well. This morning, when I opened the solution, no tests were listed in Test Explorer. (Visual Studio v15.5.6, not updated since last Friday) and I noticed many, many errors in the Output window (for Tests). They all look similar to this:

```
[12.02.2018 16:19:18 Warning] Exception System.InvalidOperationException, Exception converting Blah.Collections.BlahBlahTests.Blah: Equality: test name here
[12.02.2018 16:19:18 Warning] Operation is not valid due to the current state of the object.
[12.02.2018 16:19:18 Warning]    at Mono.Cecil.ModuleDefinition.ProcessDebugHeader()
   at Mono.Cecil.ModuleDefinition.ReadSymbols(ISymbolReader reader)
   at Mono.Cecil.ModuleReader.ReadSymbols(ModuleDefinition module, ReaderParameters parameters)
   at Mono.Cecil.ModuleReader.CreateModuleFrom(Image image, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(Stream stream, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at NUnit.VisualStudio.TestAdapter.NavigationDataProvider.CacheNewTypes(String assemblyPath, IDictionary`2 types)
   at NUnit.VisualStudio.TestAdapter.NavigationDataProvider.GetNavigationData(String className, String methodName)
   at NUnit.VisualStudio.TestAdapter.TestConverter.MakeTestCaseFromXmlNode(XmlNode testNode)
   at NUnit.VisualStudio.TestAdapter.TestConverter.ConvertTestCase(XmlNode testNode)
   at NUnit.VisualStudio.TestAdapter.NUnit3TestDiscoverer.ProcessTestCases(XmlNode topNode, ITestCaseDiscoverySink discoverySink, TestConverter testConverter)
```

One such stack trace is listed for every single test in my test project.  All other projects get the expected `[12.02.2018 16:19:17 Informational] Assembly contains no NUnit 3.0 tests...` message.

I have tried all the usual work-arounds for fixing such bugs:

1. Deleting `.vs`, `bin` and `obj` directories.
2. A clean clone of the code-base (this *did* actually work, precisely once, but the problem recurred and this work-around didn't help a second time)
3. Restart.
4. Checking for VS updates.
5. Using the Visual Studio test-runner extension and not the NUget package.
6. Referencing nunit.framework.dll from a path and not via NUget packages.
6. Checking the test runner architecture is still `x64` (it often reverts to `x86`) although this isn't really relevant for the solution in question, which is built for Any CPU.
7. Windows Update?? (Hey, I've seen it break stuff.)

At the moment, my show can be considered stopped.

For information purposes:

- I am using the new C# project file format but I have been using it for a while and this problem is completely new, as of this morning.
- My test class library is .NET Framework 4.7
- Other class libraries in the solution are .NET Standard 2.0 and they are referenced in the test class library.
- No native dependencies are in use and all code is built for "Any CPU" -- this is an upstream solution for several of my projects.

**Comment by stephenmartindale** [comment]
I have now tried changing all projects to .NET 4.7.1, thinking that the netstandard projects could be the cause. This has not resolved the problem.

**Comment by jnm2** [comment]
This looks like a duplicate of https://github.com/nunit/nunit3-vs-adapter/issues/296.
Here's the workaround: https://github.com/nunit/nunit3-vs-adapter/issues/296#issuecomment-279354920

In other news, even if that workaround doesn't work, I started the removal of Mono.Cecil a few days ago and I'll most likely be able to finish all the work needed for the fix tonight.

**Comment by jnm2** [comment]
@OsirisTerje As soon as we get https://github.com/nunit/nunit3-vs-adapter/pull/450 merged, I rebase the Mono.Cecil removal on it and we merge that PR, how soon do you think we might be able to release 3.10?

**Comment by jnm2** [comment]
Won't make it tonight. Should be this week though.

**Comment by stephenmartindale** [comment]
This work-around did not work for me. I will wait until `Mono.Cecil` is removed and try again, perhaps my bug will be fixed by collateral damage.

**Comment by jnm2** [comment]
The work is done but merging is [blocked on a preexisting issue](https://github.com/nunit/nunit3-vs-adapter/issues/296#issuecomment-365775631) and right now I don't know if that'll be resolved any time soon. I'll certainly put my focus on that as an extension of https://github.com/nunit/nunit3-vs-adapter/issues/296 but I'll need team input.

That call stack is no longer possible with the changes I've made, so I'm confident that you're right about the collateral damage. If you like I can get you a build of https://github.com/nunit/nunit3-vs-adapter/tree/jnm2/replace_cecil. It's finished and working properly.

**Comment by CharliePoole** [comment]
@jnm2 Is this an adapter issue or is there some NUnit part that also has to be fixed?

**Comment by jnm2** [comment]
This is just an adapter issue. I was going to just close it as a duplicate. Should we move it?

**Comment by CharliePoole** [comment]
IMO you should close if you think it's a duplicate, or move it to the adapter if you want@OsirisTerje to decide.

**Comment by jnm2** [comment]
I'm confused by the fact that it's not a duplicate and indeed looks like a separate interesting issue, but one which is fixed by the same work. Close enough I suppose. :D

**Comment by jnm2** [comment]
ICYMI: https://github.com/nunit/nunit3-vs-adapter/issues/296#issuecomment-365827649 🎉

**Comment by jnm2** [comment]
ICYMI: [Everything is done except for waiting for reviews on three PRs!](https://github.com/nunit/nunit3-vs-adapter/issues/296#issuecomment-365827649) 🎉
